src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/LengthBuckets.cs

```diff
@@ -87,6 +102,70 @@ internal static class LengthBuckets
                 return null;
             }
 
+#if NET5_0_OR_GREATER
+            Span<int> sortingBucket = stackalloc int[MaxPerLength];
+#else
+            int[] sortingBucket = new int[MaxPerLength];
+#endif
+            IndexedKeyComparer indexedKeyComparer = new IndexedKeyComparer(
+                comparer == EqualityComparer<string>.Default
+                ? StringComparer.Ordinal
+                : (StringComparer)comparer, keys);
+            for (int bucketStartIndex = 0; bucketStartIndex < arraySize; bucketStartIndex += MaxPerLength)
+            {
+                // Within the bucket we order the elements to make binary search easier.
+                // The first element is the first to be compared and thus is the middle value.
+                // e.g. 0,1,2,3,4,5,6 becomes 3,1,0,2,5,4,6
```

Comment:
If the first element should be in the middle, why the middle value in this sample is 2 and not 0? 
Shouldn't it be 5, 3, 1, 0, 2, 4, 6?

Reply:
In this PR I used Pre-Order traversal of the binary tree. `3,1,0,2,5,4,6` is the pre-order traversal for the tree below.

1. Looking at the slice `[..]` aka `[0..7]` We compare with the first element (`3`) to choose which branch to visit next.
2. If our number is equal to `3` then success.
3. Otherwise
   - if our number is lower than `3` then repeat using slice `[1..4]` = `102` and compare with first element (`1`).
   - if our number is higher than `3` then repeat using slice `[4..7]` = `546` and compare with first element (`5`)
```
      3
    /   \
  1      5
 / \    / \
0   2  4   6
```

Using pre-order traversal, we always move forwards in the array and the root of the tree is always the first element.